### PR TITLE
Anomaly likelihood py

### LINKED ISSUE #455: Provide missing AnomalyLikelihood from python
as our cpp Likelihood is likely broken and untested, #124 

- Python Likelihood would go to `py/` where pure python code resides. 
- it should be easy to port as has very few requirements
```

import collections
import math
import numbers
import numpy
from nupic.serializable import Serializable
from nupic.utils import MovingAverage
try:
import capnp
except ImportError:
capnp = None
if capnp:
from nupic.algorithms.anomaly_likelihood_capnp import AnomalyLikelihoodProto
class AnomalyLikelihood(Serializable):
```
   - [ ] we need either: py bindings for MovingAverage (and port py tests) 
   - [ ] replace MovingAverage in py likelihood with python code for it. (maybe prefered?) 

This will be used for running NAB #205

### PR DESCRIPTION
**Pull request by breznak**
- [x] `AnomalyLikelihood` available from pure-python code (/py/) + tests passing
  - [x] this pulls `utils.py (MovingAverage)` + tests
  - [x] `nupic.support` for unit-tests
- [x] some python 2-3 compatibility fixes
- [x] removed canpn serialization related stuff (will not work in this repo) 

Fixes #455 , related #124 #205

### CODE REVIEW

**Comment by ctrl-z-9000-times** [comment]
Wow, this is a lot of code!  This will take some time to review...

File: py/src/nupic/algorithms/anomaly_likelihood.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+               learningPeriod=288,
+               estimationSamples=100,
+               historicWindowSize=8640,
+               reestimationPeriod=100):
+    """
+    NOTE: Anomaly likelihood scores are reported at a flat 0.5 for
+    learningPeriod + estimationSamples iterations.
+
+    claLearningPeriod and learningPeriod are specifying the same variable,
+    although claLearningPeriod is a deprecated name for it.
```
Comment (ctrl-z-9000-times) [inline py/src/nupic/algorithms/anomaly_likelihood.py]: We should go ahead and remove this deprecated thing.

File: py/src/nupic/algorithms/anomaly_likelihood.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+        anomalyProbability = anomalyLikelihood.anomalyProbability(
+            value, anomalyScore, timestamp)
+
+  """
+
+
+  def __init__(self,
+               claLearningPeriod=None,
+               learningPeriod=288,
+               estimationSamples=100,
```
Comment (ctrl-z-9000-times) [inline py/src/nupic/algorithms/anomaly_likelihood.py]: Arguments learningPeriod & estimationSamples do very similar things, can they be merged?

File: py/src/nupic/algorithms/anomaly_likelihood.py [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+  s = [r[2] for r in aggRecordList]
+  dataValues = numpy.array(s)
+
+  # Estimate the distribution of anomaly scores based on aggregated records
+  if len(aggRecordList) <= skipRecords:
+    distributionParams = nullDistribution(verbosity = verbosity)
+  else:
+    distributionParams = estimateNormal(dataValues[skipRecords:])
+
+    # HACK ALERT! The HTMPredictionModel currently does not handle constant
```
Comment (ctrl-z-9000-times) [inline py/src/nupic/algorithms/anomaly_likelihood.py]: !

File: py/src/nupic/algorithms/anomaly_likelihood.py [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+predictability. For example, a likelihood of 0.01 or 1% means we see this much
+predictability about one out of every 100 records. The number is not as unusual
+as it seems. For records that arrive every minute, this means once every hour
+and 40 minutes. A likelihood of 0.0001 or 0.01% means we see it once out of
+10,000 records, or about once every 7 days.
+
+USAGE
++++++
+
+There are two ways to use the code: using the
```
Comment (ctrl-z-9000-times) [inline py/src/nupic/algorithms/anomaly_likelihood.py]: Is this code actually complicated enough to justify two APIs?

File: py/src/nupic/algorithms/anomaly_likelihood.py [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+
+    if verbosity > 2:
+      print(("Aggregating input record:", record))
+      print(("Result:", [record[0], record[1], avg]))
+
+  return averagedRecordList, historicalValues, total
+
+
+
+def estimateNormal(sampleData, performLowerBoundCheck=True):
```
Comment (ctrl-z-9000-times) [inline py/src/nupic/algorithms/anomaly_likelihood.py]: I'm pretty sure that numpy/scipy have library functions to do this, which return objects with all sorts of conveniences attached.

**Comment by breznak** [comment]
> This will take some time to review... 

Actually I hope this would be an easy review :) 

It's basically 1:1 to numenta/nupic's code, which I take for review. 
I suggest reviewing commit by commit to see only the changes I did. Rest is pulled from nupic.py branch, which tracks nupic.py repo, which tracks original,old numenta/nupic (with some fixes). 

I just checked out the needed files, moved them around to correct locations, and made minimal fixes for py2/3,capnp.

File: py/src/nupic/algorithms/anomaly_likelihood.py (hunk 1 above)
Comment (breznak) [reply py/src/nupic/algorithms/anomaly_likelihood.py]: I'd prefer not in this PR, this just pulls the code from python to this repo.

**Comment by breznak** [comment]
> ./../../../build/Release/distr/src/nupic/support/__init__.py:46: in <module>
    from .configuration import Configuration
E   ImportError: No module named configuration 

I was wondering why that one is not needed locally.

**Comment by breznak** [comment]
This PR is large for need to bring up support for runing py unittests from nupic. Once it's inplace, porting a new test should be as simple as just copying the file.

File: py/src/nupic/algorithms/anomaly_likelihood.py (hunk 5 above)
Comment (breznak) [reply py/src/nupic/algorithms/anomaly_likelihood.py]: First of all, I want to say I really appreciate you do deep code review where you understand the code and question the implementation! :+1: 

As for many of these comments, I agree, but I don't want to act on them in this PR which just brings 1:1 code from numenta's, which can be considered for sort of a standard. 

> I'm pretty sure that numpy/scipy have library functions to do this, which return objects with all sorts of conveniences attached.

as for the justified changes. Yes, the code could (should?) be improved. The result of your review should be fixed in nupic.py (and then brought back here?) 

Q: do we want to deviate from Numenta's python implementation? (we said (only API-wise?) we'll keep strong compatibility. 

Q: the nupic repo is definitely more alive than nupic.core, should we offer fix there? @rhyolight will you accept it? 

Q: actually, do we want to (long term) have PY code here, shouldn't we just focus on API compat with numenta/nupic or community/nupic.py? (I think it's fine to merge all py code in this repo)

Comment (breznak) [reply py/src/nupic/algorithms/anomaly_likelihood.py]: see https://github.com/numenta/nupic/issues/3876

**Comment by ctrl-z-9000-times** [comment]
>  numenta/nupic's code, which I take for reviewed.

I am deeply suspicious of numenta's code.  Their ideas are great, their implementations are in places half baked.  However since we have a clear need for this code, and because i dont actually have the time to make this better, i will go easy on this code review and not demand too many changes.

**Comment by breznak** [comment]
> I am deeply suspicious of numenta's code. Their ideas are great, their implementations are in places half baked.

I have to agree on this one. 
Still, that code is published in peer-reviewed papers. It'll be good to have it as baseline, if we fix bugs or make it better, it'll be a documented and replicable improvement. 

I'll open a follow up issue to have a crack at AnomalyLikelihood code review when we have time.

File: py/src/nupic/tests/algorithms/anomaly_likelihood_jeff_test.py [hunk 6]
```diff
@@ -35,7 +35,7 @@
 import unittest
 
 from nupic.algorithms import anomaly_likelihood as an
-from nupic.support.unittesthelpers.testcasebase import TestCaseBase
+from unittest import TestCase as TestCaseBase
```
Comment (breznak) [inline py/src/nupic/tests/algorithms/anomaly_likelihood_jeff_test.py]: I like this change! Makes the diff much smaller, and I agree we should limit the scope of things we care about. 
Thanks :+1:

**Pull request merged**